compiler/src/main/java/com/squareup/anvil/compiler/codegen/CodeGenerationExtension.kt

```diff
@@ -0,0 +1,97 @@
+package com.squareup.anvil.compiler.codegen
+
+import com.squareup.anvil.compiler.codegen.CodeGenerator.GeneratedFile
+import org.jetbrains.kotlin.analyzer.AnalysisResult
+import org.jetbrains.kotlin.analyzer.AnalysisResult.RetryWithAdditionalRoots
+import org.jetbrains.kotlin.com.intellij.openapi.project.Project
+import org.jetbrains.kotlin.com.intellij.psi.PsiManager
+import org.jetbrains.kotlin.com.intellij.testFramework.LightVirtualFile
+import org.jetbrains.kotlin.container.ComponentProvider
+import org.jetbrains.kotlin.context.ProjectContext
+import org.jetbrains.kotlin.descriptors.ModuleDescriptor
+import org.jetbrains.kotlin.idea.KotlinFileType
+import org.jetbrains.kotlin.psi.KtFile
+import org.jetbrains.kotlin.resolve.BindingTrace
+import org.jetbrains.kotlin.resolve.jvm.extensions.AnalysisHandlerExtension
+import java.io.File
+
+internal class CodeGenerationExtension(
+  private val codeGenDir: File
+) : AnalysisHandlerExtension {
+
+  private var didRecompile = false
+
+  private val codeGenerators = mutableListOf<CodeGenerator>()
+
+  fun registerCodeGenerator(codeGenerator: CodeGenerator) {
+    codeGenerators += codeGenerator
+  }
+
+  fun unregisterCodeGenerator(codeGenerator: CodeGenerator) {
+    codeGenerators -= codeGenerator
+  }
+
+  override fun doAnalysis(
+    project: Project,
+    module: ModuleDescriptor,
+    projectContext: ProjectContext,
+    files: Collection<KtFile>,
+    bindingTrace: BindingTrace,
+    componentProvider: ComponentProvider
+  ): AnalysisResult? {
+    // Tell the compiler that we have something to do in the analysisCompleted() method if
+    // necessary.
+    return if (!didRecompile && codeGenerators.isNotEmpty()) AnalysisResult.EMPTY else null
+  }
+
+  override fun analysisCompleted(
+    project: Project,
+    module: ModuleDescriptor,
+    bindingTrace: BindingTrace,
+    files: Collection<KtFile>
+  ): AnalysisResult? {
+    if (didRecompile || codeGenerators.isEmpty()) return null
+    didRecompile = true
+
+    codeGenDir.listFiles()
+        ?.forEach {
+          check(it.deleteRecursively()) {
+            "Could not clean file: $it"
+          }
+        }
+
+    fun generateCode(files: Collection<KtFile>): Collection<GeneratedFile> =
+      codeGenerators
+          .map {
+            it.generateCode(codeGenDir, module, files)
+          }
+          .flatten()
+
+    val psiManager = PsiManager.getInstance(project)
+
+    var newFiles = generateCode(files)
+    while (newFiles.isNotEmpty()) {
```

Comment:
There's never a scenario where you'd end up looping forever due to the newFiles list never being empty? I'm slightly worried that you'd run into an issue where you'd need to use an iterator instead. Just making sure.

Reply:
No, that doesn't happen.